[최용석] Step2 PR 리뷰 요청드립니다.

badaelephant opened the pull request:
안녕하세요 장민제 리뷰어님, 스텝 2 PR 리뷰 요청 드립니다. 

이번에는 outside in 방식으로 구현을 했는데요. 

작업하면서 아직 outside-in 방식보다는, 로직이 복잡할 수록 더 inside-out 방식으로 접근하는게 개인적으론 편하네요. 

좀 더 공부하고 노력해 봐야 할 것 같습니다. 

이번에도 리뷰 잘 부탁드립니다. ^^

Commit:
docs: 스텝2 요구사항 정리

Commit:
refactor: 정적 팩토리 메서드 추가

Commit:
test: 경로조회 인수테스트 작성

Commit:
feat 경로조회 컨트롤러 개발

Commit:
test: 경로조회 서비스 테스트 구현

Commit:
feat: 경로조회 서비스 기능 구현

Commit:
feat: 경로 응답객체 및 도메인 객체 생성

Commit:
feat: 경로조회 도메인 객체 구현, interface 구현

Commit:
test: 경로조회 단위테스트 구현하여 도메인 동작검증

Commit:
refactor: 마이너 변경 - unmodifiableList 추가, 불필요한 콘솔출력 삭제, import 정리 등

minzzang commented on README.md:
```diff
@@ -36,4 +36,18 @@ Feature: 지하철 구간 관련 기능
+
+### 요구사항
+
+- [x] 최단 경로 조회 인수 테스트 만들기
```
요구사항 정리 👍

badaelephant replied:
감사합니다.~

minzzang commented on src/main/java/nextstep/subway/path/domain/DijkstraShortestPathStrategy.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public void createGraphPath(Graph graph) {
+        this.path = new DijkstraShortestPath(graph);
```
createGraphPath를 호출하기 전에 getGraphPath를 호출하면 예외가 발생할 수 있을 것 같은데 초기화를 나중에 해주신 이유가 있을까요~?

badaelephant replied:
아 이부분은 제가 의존성 주입을 잘못하면서 그렇게 된거라 수정하였습니다. 

이제는 StationGraph <-- PathFinder로 단일 의존성을 갖도록 수정하엿습니다. 
이전에는 잘못 했던것이 StationGraph<-->PathFindStrategy가 양쪽에 서로 의존성을 갖고있었어요. 잘못 만든것 같습니다.

minzzang commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,29 @@
+import nextstep.subway.line.domain.Distance;
+import nextstep.subway.station.domain.Station;
+
+public class Path {
```
Path 도메인 분리 👍

badaelephant replied:
감사합니다.ㅎㅎ

minzzang commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,28 @@
+        return new PathFinder(stationGraph);
+    }
+
+    public Path findPathFromGraph(Station source, Station target) {
```
stationGraph.findPath만 호출하고 있는데 PathFinder를 따로 분리해주신 이유가 있을까요~?

badaelephant replied:
이부분도 수정해서 stationGraph는 Graph의 생성, PathFinder는 경로 탐색으로 역할을 구분하였습니다.

minzzang commented on src/main/java/nextstep/subway/path/domain/PathStrategy.java:
```diff
@@ -0,0 +1,12 @@
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+
+public interface PathStrategy {
```
전략 패턴 👍

badaelephant replied:
감사합니다. ㅎㅎ

minzzang commented on src/main/java/nextstep/subway/path/domain/StationGraph.java:
```diff
@@ -0,0 +1,59 @@
+        GraphPath graphPath = pathStrategy.getGraphPath(source, target);
+        validateConnected(graphPath);
+        List<Station> shortestPath = Collections.unmodifiableList(graphPath.getVertexList());
+        Double distance = pathStrategy.getGraphPath(source, target).getWeight();
```
`graphPath.getWeight();`로 변경할 수 있을 것 같네요! 😄
그리고 줄 바꿈을 해준다면 조금 더 가독성 있는 코드가 될 것 같아요!

badaelephant replied:
수정하였습니다. ㅎㅎ

minzzang commented on src/test/java/nextstep/subway/path/application/PathFinderServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+
+    @DisplayName("최단거리 경로 조회")
+    @Test
+    void findShortestPath() {
```
mock을 활용하여 service test 잘 만들어 주셨네요! 👍 
inside-out방식과 비교했을 때 어떤 장단점이 있으셨나요~?
또한 mock을 활용했을 때 어떤 불편함이 있으셨을까요? mock객체 대신에 fake객체를 활용하셔도 좋을 것 같아요~!

badaelephant replied:
outside in 방식의 경우 사실 제가 잘 활용을 못해서 그런걸지는 몰라도.. 아무래도, 저희 nextstep의 미션 같은 경우는, 대부분 도메인 로직이 상당히 머리를 써야 하는 경우가 많은 것 같은데, 도메인을 어떻게 구현할지 / 어떤식으로 전략패턴을 가져갈지, 도메인 객체의 역할은 구체적으로 어떻게 나눌 지 등에 대해서 큰 도움이 되는 느낌을 못받았습니다. 

지금 미션처럼 경로조회라는 새로운 기능을 만드는 것이 단일 기능이 아니라 다른 서비스- 기능들과 연관관계가 있을때는 연결부를 명확하게 하고 도메인 기능을 개발한다면 더 효용성이 높을 것같습니다. 

저는 Mock을 활용하면서. 제가 아직도 이 방식의 테스트에 익숙하지 않아서인지 한가지가 크게 어려웠는데요
대부분의 entity 생성자가 id가 없는데, mock 객체에 id가 없어서, 뒤에 테스트에서 null point exception이 발생하는 경우가 종종 있었습니다. => mock 객체에 id를 넣어줄 수 있는 방법이 있을까요?

minzzang replied:
필요에 따라 생성자를 활용하여 생성할 수 있고 ReflectionTestUtils을 활용하여 id를 넣어줄 수 있습니다~ 😄

minzzang commented on src/test/java/nextstep/subway/path/unit/StationGraphTest.java:
```diff
@@ -0,0 +1,116 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class StationGraphTest {
```
꼼꼼한 단위 테스트 👍

badaelephant replied:
감사합니다.

minzzang commented on src/test/java/nextstep/subway/path/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,155 @@
+     * When 두개 역 사이의 최단경로를 조회하면
+     * Then 최단 경로 거리, 최단 경로를 확인할 수 있다.
+     */
+    @DisplayName("두 역 사이의 최단경로를 조회할 수 있다.")
```
인수 테스트 잘 작성해주셨네요~! 👍 

예외 케이스까지 꼼꼼하게 작성해주셨는데 StationGraph 단위 테스트에서 해당 예외를 테스트하고 있어서 테스트 중복이라고 생각을 하실 수 있을 것 같은데 

용석님은 인수 테스트와 단위 테스트를 어떤 기준을 가지고 작성을 하고 계실까요~?

badaelephant replied:
확실히 말씀하신대로 예외처리를 중복적으로 하지 않기 위해서는 단위테스트에 예외검증을 위임하고, 인수테스트는 해피케이스 위주로, 그리고 혹시 실패하는 경우를 검증하기 위해선 통합테스트에서 해당 시나리오를 검증하는 것이 좋아 보입니다. 이부분 수정하였습니다.

minzzang replied:
말씀해주신 통합 테스트에서 실패하는 경우를 검증하여 더 안정적인 코드를 만들고 개발자가 테스트로 인해 안정감을 얻는 방법도 좋아보이네요 👍

minzzang reviewed (changes requested):
안녕하세요 용석님~!

경로 조회 미션 역시 잘 작성해주셨네요! :+1:

소소한 코멘트와 질문을 남겼는데 확인 부탁드립니다. :pray:

남은 주말도 화이팅입니다. :muscle:

Commit:
refactor : 피드백 반영
1. 경로조회와 관련된 도메인 객체의 역할 재정의
PathFinder : Graph를 인자로 받아서 생성, 시작역-종료역을 받아서 Path를 리턴함.
StationGraph : Section 리스트를 받아서 Graph를 리턴함.

2. PathFinder의 생성자 정비

3. 예외검증은 단위테스트에서, 인수테스트는 Happy Case위주로 변경

minzzang commented on src/main/java/nextstep/subway/path/domain/DijkstraShortestPathFinder.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public Path findPath(Station source, Station target) {
```
로직이 깔끔하네요! 😄

Commit:
refactor: unmodifiablelist 추가

minzzang reviewed (approved):
안녕하세요 용석님~!

피드백 반영 잘해주셨네요! :+1:

다음 미션도 화이팅입니다. :muscle:
